### PR TITLE
Remove the Google web fonts field from the page layout

### LINKED ISSUE #1791: Google-Fonts API Version 2 Handling
As mentioned in the [forum](https://community.contao.org/de/showthread.php?78134-Google-Fonts-API-Base-URL&p=526398&viewfull=1#post526398):

By using the suggested Google-Fonts embed parameters for example:
https://fonts.google.com/specimen/Oswald?sidebar.open&vfonly&query=Oswald&selection.family=Oswald:wght@200;380;400

within the Pagelayout/Webfonts settings we get unexpected results. This is because of the Google Fonts APIv2 - see: https://developers.google.com/fonts/docs/css2

So maybe this should be mentioned in the pagelayout help text. Difficult to describe ...
Or switch to the new API Version?
Or leaving a hint in the docs? 
...

### PR DESCRIPTION
**Pull request by leofeyer**
| Q                | A
| -----------------| ---
| Implemented feature     | #1791
| Docs PR or issue | -

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/Migration/Version411/MergeWebfontsMigration.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        $headFieldQuoted = $this->connection->quoteIdentifier('head');
+
+        $rows = $this->connection->fetchAllAssociative("
+            SELECT
+                id, $webfontsFieldQuoted, $headFieldQuoted
+            FROM
+                $tableQuoted
+            WHERE
+                $webfontsFieldQuoted != ''
+        ");
```
Comment (m-vo) [inline core-bundle/src/Migration/Version411/MergeWebfontsMigration.php]:
```suggestion
        $rows = $this->connection->fetchAllAssociative("SELECT id, webfonts, head FROM tl_layout WHERE webfonts != ''");
```

File: core-bundle/src/Migration/Version411/MergeWebfontsMigration.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            $this->connection
+                ->prepare("UPDATE $tableQuoted SET $headFieldQuoted = :head WHERE id = :id")
+                ->execute([
+                    ':id' => $row['id'],
+                    ':head' => $row['head']."\n".'<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=' . str_replace('|', '%7C', $row['webfonts']) . '">',
+                ])
+            ;
+        }
+
+        $this->connection->executeStatement("ALTER TABLE tl_layout DROP $webfontsFieldQuoted");
```
Comment (m-vo) [inline core-bundle/src/Migration/Version411/MergeWebfontsMigration.php]:
```suggestion
        $this->connection->executeStatement('ALTER TABLE tl_layout DROP webfonts');
```

**Review by m-vo — CHANGES_REQUESTED** [review]

File: core-bundle/src/Migration/Version411/MergeWebfontsMigration.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/Migration/Version411/MergeWebfontsMigration.php]: Not sure about this. I took the code from @ausi's migration:

https://github.com/contao/contao/blob/b0f4b53f4e869d63af16323aa24a4d1bd6adc421/core-bundle/src/Migration/Version410/OrderFieldMigration.php#L98-L112

Comment (m-vo) [reply core-bundle/src/Migration/Version411/MergeWebfontsMigration.php]: Mh, I don't like this added complexity - this also breaks SQL validation in PhpStorm.

If you still want to go down this route, there is a leftover `tl_layout` in the last query. :stuck_out_tongue:

File: core-bundle/src/Migration/Version411/MergeWebfontsMigration.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                id, $webfontsFieldQuoted, $headFieldQuoted
+            FROM
+                $tableQuoted
+            WHERE
+                $webfontsFieldQuoted != ''
+        ");
+
+        foreach ($rows as $row) {
+            $this->connection
+                ->prepare("UPDATE $tableQuoted SET $headFieldQuoted = :head WHERE id = :id")
```
[the remark this answers is not in the document]
Comment (m-vo) [reply core-bundle/src/Migration/Version411/MergeWebfontsMigration.php:68]: Might even want to change this to `$this->connection->update(...)`.

File: core-bundle/src/Migration/Version411/MergeWebfontsMigration.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                ->execute([
+                    ':id' => $row['id'],
+                    ':head' => $row['head']."\n".'<link rel="stylesheet" href="https://fonts.googleapis.com/css?family='.str_replace('|', '%7C', $row['webfonts']).'">',
+                ])
+            ;
```
Comment (m-vo) [inline core-bundle/src/Migration/Version411/MergeWebfontsMigration.php]:
```suggestion
            $head = $row['head']."\n".'<link rel="stylesheet" href="https://fonts.googleapis.com/css?family='.str_replace('|', '%7C', $row['webfonts']).'">';

            $this->connection->update('tl_layout', ['head' => $head], ['id' => $row['id']]);
```

File: core-bundle/src/Migration/Version411/MergeWebfontsMigration.php (hunk 1 above)
Comment (ausi) [reply core-bundle/src/Migration/Version411/MergeWebfontsMigration.php]: `quoteIdentifier()` is only needed when the identifier is dynamic. For static identifiers like in this case we normaly write them directly into the query as suggested by @m-vo

Comment (leofeyer) [reply core-bundle/src/Migration/Version411/MergeWebfontsMigration.php]: Changed in 7e3c42d2bc2592001430c275161e5fe5e3b75bd6.

File: core-bundle/src/Migration/Version411/MergeWebfontsMigration.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Migration/Version411/MergeWebfontsMigration.php]: Changed in 7e3c42d2bc2592001430c275161e5fe5e3b75bd6.

**Review by aschempp — APPROVED** [review]
I would check if there's any layout with fonts in `shouldRun` but otherwise looks good

**Comment by leofeyer** [comment]
The migration needs to run even if there are no layouts with web fonts, because it will remove the database column and thus prevent itself from running multiple times.

**Comment by aschempp** [comment]
> The migration needs to run even if there are no layouts with web fonts, because it will remove the database column and thus prevent itself from running multiple times.

well if there's no field with data, it does not need to run, and it wouldn't run again because there's still no field with data 😉

**Comment by frontendschlampe** [comment]
I think, to migrate the field isn't a good idea. I would give a hint to anybody, that the field will be deleted in Contao 4.11 and the webmaster should add the fonts again (or change it before update). If you update from 4.10 to 4.11, you have to do and check much more than just pressing the update button anyway. Maybe a hint on homescreen in the backend ...

**Comment by aschempp** [comment]
@frontendschlampe not sure what you're referring to. Migrations are there to make sure people can upgrade without breaking their sites. You're suggesting we should break their site, which makes no sense except something like an "opinion" they _should_ switch to local fonts…

**Comment by frontendschlampe** [comment]
First of all: we don't break any installation because of a webfont. Then it's an update which will surely require more adjustments than just pressing the update button in the Manager. In addition, third party extensions that might do something with the web fonts will get function problems because the field is no longer available or the desired function of the extension is no longer available - and this without any message. Not to mention the recommendation for local integration of web fonts.

But it's just my opinion. We have lot of other things to do, so you don't have to worry about copying web fonts. It's everytime the same discussion: should we migrate or not.

**Comment by m-vo** [comment]
>  Then it's an update which will surely require more adjustments than just pressing the update button in the Manager

> We have lot of other things to do

I don't think so. After running the migrations you should be fine. Where are you seeing problems?

**Comment by frontendschlampe** [comment]
> I don't think so. After running the migrations you should be fine. Where are you seeing problems?

We talk about 4.11 ... I don't know. We will see ...

> We have lot of other things to do

This sentence does not refer to an update, but to the load on the core team. ;-)

**Pull request merged**